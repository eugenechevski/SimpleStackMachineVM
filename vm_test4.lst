Address Instruction
     0: NOTR 
     1: STRA 
     2: ADDI $sp, 0, 5
     3: ADDI $sp, -1, 2
     4: ARI $sp, -1
     5: ADD $sp, 0, $sp, 1
     6: MUL $sp, 1
     7: SRI $sp, 1
     8: CFLO $sp, 0
     9: SUB $sp, 0, $sp, 1
    10: MUL $sp, 2
    11: SRI $sp, 1
    12: CFLO $sp, 0
    13: SRI $sp, 1
    14: CPW $sp, 0, $sp, 2
    15: DIV $sp, 4
    16: SRI $sp, 2
    17: CFLO $sp, 0
    18: CFHI $sp, 1
    19: SRI $sp, 1
    20: ADDI $sp, 0, 5
    21: BEQ $sp, 1, 2	# target is word address 23
    22: CALL 35	# target is word address 35
    23: CALL 31	# target is word address 31
    24: ARI $sp, 1
    25: SWR $sp, 0, $r4
    26: ADDI $sp, 0, 3
    27: BEQ $sp, 1, 2	# target is word address 29
    28: CALL 35	# target is word address 35
    29: CALL 31	# target is word address 31
    30: EXIT 0
    31: NOTR 
    32: PSTR $gp, 0
    33: STRA 
    34: RTN 
    35: NOTR 
    36: PSTR $gp, 1
    37: STRA 
    38: EXIT 1
    39: RTN 
     512: 2681	     513: 683886	     514: 0	        ...     

