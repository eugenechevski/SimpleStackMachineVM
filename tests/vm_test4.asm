# $Id: vm_test4.asm,v 1.5 2024/08/29 20:26:00 leavens Exp $
	.text start
start:	NOTR
	STRA
	ADDI $sp, 0, 5     # stack top is 5
	ADDI $sp, -1, 2    # memory[SP-1] is 2
	ARI $sp, -1        # make stack top point to where the 2 is
	ADD $sp, 0, $sp, 1 # stack top should be 7
	MUL $sp, 1         # LO should be (5 * 7) = 35
	SRI $sp, 1         # allocate a word on stack
	CFLO $sp, 0        # stack top is 35
	SUB $sp, 0, $sp, 1 # stack top should be (35-7) = 28
	MUL $sp, 2         # LO should be (28 * 5) = 140
        SRI $sp, 1         # allocate a word on stack
	CFLO $sp, 0        # stack top is 140
	SRI $sp, 1         # allocate a word on stack
	CPW $sp, 0, $sp, 2 # put 28 on top of stack
	DIV $sp, 4	   # HI should be (28 % 5) = 3, LO s.b. (28 / 5) = 5
        SRI $sp, 2         # allocate two words on stack
	CFLO $sp, 0        # stack top should be 5
	CFHI $sp, 1        # second from top should be 3
	SRI $sp, 1         # allocate a word on the stack
	ADDI $sp, 0, 5     # store 5! on top of stack
	BEQ $sp, 1, 2      # write no and exit 1 if memory[$sp+1] is not 5
	CALL no
	CALL yes
	ARI $sp, 1         # pop a word off the stack
	SWR $sp, 0, $r4    # store a 0 (from $r4) into top of stack
	ADDI $sp, 0, 3     # make top of stack be 3
	BEQ $sp, 1, 2      # write no if memory[$sp+1] is not 3
	CALL no
	CALL yes
done:	EXIT 0

yes:	NOTR
	PSTR $gp, 0       # print "y\n"
	STRA
	RTN

no:	NOTR
	PSTR $gp, 1       # print "no\n"
	STRA
	EXIT 1
        RTN
	
	.data 512
	STRING[1] y = "y\n"
	STRING[1] n = "no\n"
	.stack 4096
	.end
	
