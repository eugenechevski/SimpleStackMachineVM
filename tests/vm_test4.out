      PC: 0
GPR[$gp]: 512  	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4096: 0	

==>      0: NOTR 
      PC: 2
GPR[$gp]: 512  	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4096: 0	

==>      2: ADDI $sp, 0, 5
      PC: 3
GPR[$gp]: 512  	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4096: 5	

==>      3: ADDI $sp, -1, 2
      PC: 4
GPR[$gp]: 512  	GPR[$sp]: 4096 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     
    4095: 2	
    4096: 5	

==>      4: ARI $sp, -1
      PC: 5
GPR[$gp]: 512  	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4095: 2	    4096: 5	

==>      5: ADD $sp, 0, $sp, 1
      PC: 6
GPR[$gp]: 512  	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4095: 7	    4096: 5	

==>      6: MUL $sp, 1
      PC: 7	      HI: 0	      LO: 35
GPR[$gp]: 512  	GPR[$sp]: 4095 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4095: 7	    4096: 5	

==>      7: SRI $sp, 1
      PC: 8	      HI: 0	      LO: 35
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4094: 0	    4095: 7	    4096: 5	

==>      8: CFLO $sp, 0
      PC: 9	      HI: 0	      LO: 35
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4094: 35	    4095: 7	    4096: 5	

==>      9: SUB $sp, 0, $sp, 1
      PC: 10	      HI: 0	      LO: 35
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4094: 28	    4095: 7	    4096: 5	

==>     10: MUL $sp, 2
      PC: 11	      HI: 0	      LO: 140
GPR[$gp]: 512  	GPR[$sp]: 4094 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4094: 28	    4095: 7	    4096: 5	

==>     11: SRI $sp, 1
      PC: 12	      HI: 0	      LO: 140
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4093: 0	    4094: 28	    4095: 7	    4096: 5	

==>     12: CFLO $sp, 0
      PC: 13	      HI: 0	      LO: 140
GPR[$gp]: 512  	GPR[$sp]: 4093 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4093: 140	    4094: 28	    4095: 7	    4096: 5	

==>     13: SRI $sp, 1
      PC: 14	      HI: 0	      LO: 140
GPR[$gp]: 512  	GPR[$sp]: 4092 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4092: 0	    4093: 140	    4094: 28	    4095: 7	    4096: 5	

==>     14: CPW $sp, 0, $sp, 2
      PC: 15	      HI: 0	      LO: 140
GPR[$gp]: 512  	GPR[$sp]: 4092 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4092: 28	    4093: 140	    4094: 28	    4095: 7	    4096: 5	

==>     15: DIV $sp, 4
      PC: 16	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4092 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4092: 28	    4093: 140	    4094: 28	    4095: 7	    4096: 5	

==>     16: SRI $sp, 2
      PC: 17	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4090: 0	        ...         4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     17: CFLO $sp, 0
      PC: 18	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4090: 5	    4091: 0	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     18: CFHI $sp, 1
      PC: 19	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4090: 5	    4091: 3	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     19: SRI $sp, 1
      PC: 20	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4089 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4089: 0	    4090: 5	    4091: 3	    4092: 28	    4093: 140	
    4094: 28	    4095: 7	    4096: 5	

==>     20: ADDI $sp, 0, 5
      PC: 21	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4089 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4089: 5	    4090: 5	    4091: 3	    4092: 28	    4093: 140	
    4094: 28	    4095: 7	    4096: 5	

==>     21: BEQ $sp, 1, 2	# target is word address 23
      PC: 23	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4089 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 0    
     512: 2681	     513: 683886	     514: 0	        ...     

    4089: 5	    4090: 5	    4091: 3	    4092: 28	    4093: 140	
    4094: 28	    4095: 7	    4096: 5	

==>     23: CALL 31	# target is word address 31
      PC: 31	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4089 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 24   
     512: 2681	     513: 683886	     514: 0	        ...     

    4089: 5	    4090: 5	    4091: 3	    4092: 28	    4093: 140	
    4094: 28	    4095: 7	    4096: 5	

==>     31: NOTR 
y
      PC: 34	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4089 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 24   
     512: 2681	     513: 683886	     514: 0	        ...     

    4089: 2	    4090: 5	    4091: 3	    4092: 28	    4093: 140	
    4094: 28	    4095: 7	    4096: 5	

==>     34: RTN 
      PC: 24	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4089 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 24   
     512: 2681	     513: 683886	     514: 0	        ...     

    4089: 2	    4090: 5	    4091: 3	    4092: 28	    4093: 140	
    4094: 28	    4095: 7	    4096: 5	

==>     24: ARI $sp, 1
      PC: 25	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 24   
     512: 2681	     513: 683886	     514: 0	        ...     
    4089: 2	
    4090: 5	    4091: 3	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     25: SWR $sp, 0, $r4
      PC: 26	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 24   
     512: 2681	     513: 683886	     514: 0	        ...     
    4089: 2	
    4090: 0	    4091: 3	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     26: ADDI $sp, 0, 3
      PC: 27	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 24   
     512: 2681	     513: 683886	     514: 0	        ...     
    4089: 2	
    4090: 3	    4091: 3	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     27: BEQ $sp, 1, 2	# target is word address 29
      PC: 29	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 24   
     512: 2681	     513: 683886	     514: 0	        ...     
    4089: 2	
    4090: 3	    4091: 3	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     29: CALL 31	# target is word address 31
      PC: 31	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 30   
     512: 2681	     513: 683886	     514: 0	        ...     
    4089: 2	
    4090: 3	    4091: 3	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     31: NOTR 
y
      PC: 34	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 30   
     512: 2681	     513: 683886	     514: 0	        ...     
    4089: 2	
    4090: 2	    4091: 3	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     34: RTN 
      PC: 30	      HI: 3	      LO: 5
GPR[$gp]: 512  	GPR[$sp]: 4090 	GPR[$fp]: 4096 	GPR[$r3]: 0    	GPR[$r4]: 0    
GPR[$r5]: 0    	GPR[$r6]: 0    	GPR[$ra]: 30   
     512: 2681	     513: 683886	     514: 0	        ...     
    4089: 2	
    4090: 2	    4091: 3	    4092: 28	    4093: 140	    4094: 28	
    4095: 7	    4096: 5	

==>     30: EXIT 0
